server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1325,6 +1328,12 @@ public void cleanupStorage(boolean recurring) {
 
                     List<VolumeVO> vols = _volsDao.listVolumesToBeDestroyed(new Date(System.currentTimeMillis() - ((long)StorageCleanupDelay.value() << 10)));
                     for (VolumeVO vol : vols) {
+                        if (Type.ROOT.equals(vol.getVolumeType()) && userVmManager.getDestroyRootVolumeOnVmDestruction(vol.getDomainId())) {
```

Comment:
@JoaoJandre 
the vm state should be checked instead of the global/domain setting, because the setting can be changed.
if vm state is in Destroyed state, skip it and continue